armi/reactor/excoreStructure.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2024 TerraPower, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module provides the simplest base-class tools for representing reactor objects that are
+outside the reactor core.
+
+The idea here is that all ex-core objects will be represented first as a spatial grid, and then
+arbitrary ArmiObjects can be added to that grid.
+"""
+import copy
+
+from armi.reactor.composites import Composite
+
+
+class ExcoreStructure(Composite):
+    """This is meant as the simplest baseclass neede to represent an ex-core reactor thing.
+
+    An ex-core structure is expected to:
+
+    - be a child of the Reactor,
+    - have a grid associated with it,
+    - contain a hierarchical set of ArmiObjects.
+    """
+
+    def __init__(self, name, parent=None):
+        Composite.__init__(self, name)
+        self.parent = parent
+        self.spatialGrid = None
+
+    def __repr__(self):
+        return "<{}: {} id:{}>".format(self.__class__.__name__, self.name, id(self))
+
+    @property
+    def r(self):
+        return self.getAncestor(fn=lambda x: x.__class__.__name__ == "Reactor")
+
+    def add(self, obj, loc=None):
```

Comment:
The `add()` method on `AssemblyList` had some functionality to allow for things to be added without specifying a location. 

https://github.com/terrapower/armi/blob/2fac7a8e2f105a29174ac33bf0e28318fc9272a3/armi/reactor/assemblyLists.py#L151-L154

It looks like you have decided to get rid of that feature. I think that was nice to have, and removing it will possibly (probably?) break some things downstream.

Reply:
You have the choice of having the `obj` contain a `obj.spatialLocator` _OR_ passing in a `loc` directly.

You can do either.

Reply:
Yes, I understand that the location can be specified in two ways. But my understanding is that the location did not used to have to be specified at all, and the assembly would then just get whatever "next position" that is open. 

That ability seemed useful, and it has now gone away.

Am I misunderstanding the new implementation?

Reply:
That logic still exists for the Spent Fuel Pool, over in `spentFuelPool.py`.

But the `ExcoreStructure` class is extremely general. These are meant to support things that people have asked me for, like the ability to define ducts and pipes on a grid.

In this, most general possible situation, all you have are "objects in space", and there is no obvious "next place in the grid".

Thoughts?

Reply:
Ah, I see. You're right, looks good to me.